app/src/main/java/com/github/se/icebreakrr/model/message/MeetingRequestService.kt

```diff
@@ -67,52 +70,66 @@ class MeetingRequestService : FirebaseMessagingService() {
+        if (!isAppInForeground()) {
+          showNotification(MSG_REQUEST, "from : $senderName")
+        }
       }
       "MEETING RESPONSE" -> {
         val accepted = remoteMessage.data["accepted"]?.toBoolean() ?: false
-
+        val message = remoteMessage.data["message"] ?: "null"
+        val locationString = remoteMessage.data["location"]?.trim('(', ')') ?: "null"
+        val latitudeString = locationString.split(", ")[0]
+        val longitudeString = locationString.split(", ")[1]
+        val location = Pair(latitudeString.toDouble(), longitudeString.toDouble())
+        val locationAndMessage = Pair(message, location)
         MeetingRequestManager.meetingRequestViewModel?.removeFromMeetingRequestSent(senderUid) {
           MeetingRequestManager.meetingRequestViewModel?.updateInboxOfMessages {}
         }
         if (accepted) {
-          showNotification(senderName + MSG_RESPONSE_ACCEPTED, "")
-          MeetingRequestManager.meetingRequestViewModel?.addPendingLocation(senderUid) {}
+          MeetingRequestManager.meetingRequestViewModel?.confirmMeetingLocation(
+              senderUid, locationAndMessage) {
+                Log.e("LOCATION CONFIRMATION", "failed to confirm the meeting location")
+              }
+          if (!isAppInForeground()) {
+            showNotification(senderName + MSG_RESPONSE_ACCEPTED, "")
+          }
         } else {
-          showNotification(senderName + MSG_RESPONSE_REJECTED, "")
+          if (!isAppInForeground()) {
+            showNotification(senderName + MSG_RESPONSE_REJECTED, "")
+          }
         }
-      }
-      "MEETING CONFIRMATION" -> {
-        val locationString = remoteMessage.data["location"] ?: "null"
-        val latitudeString = locationString.split(", ")[0]
-        val longitudeString = locationString.split(", ")[1]
-        MeetingRequestManager.meetingRequestViewModel?.confirmMeetingLocation(
-            senderUid, Pair(message, Pair(latitudeString.toDouble(), longitudeString.toDouble()))) {
-              Log.e("MeetingRequestService", "error when confirmMeetingLocation : ${it.message}")
-            }
-        showNotification(senderName + MSG_CONFIRMATION, MSG_CONFIRMATION_INFO)
+        MeetingRequestManager.meetingRequestViewModel?.stopMeetingRequestTimer(senderUid, this)
       }
       "MEETING CANCELLATION" -> {
+        val message = remoteMessage.data["message"] ?: "null"
         val stringReason =
             when (message) {
-              "distance" -> DISTANCE_REASON_CANCELLATION
+              "DISTANCE" -> DISTANCE_REASON_CANCELLATION
+              "TIME" -> TIME_REASON_CANCELLATION
               else -> DEFAULT_REASON_CANCELLATION
             }
-        showNotification("Cancelled meeting with $senderName", stringReason)
         MeetingRequestManager.meetingRequestViewModel?.removeFromMeetingRequestInbox(senderUid)
         MeetingRequestManager.meetingRequestViewModel?.removeFromMeetingRequestSent(senderUid) {}
```

Comment:
There could be racing conditions between these two functions since you update your profile on firebase two times at small intervals. 

_Suggestions_ : You can add an `onComplete()` to `removFromMeetingRequestInbox` to avoid racing conditions. You can also `stopMeetingRequestTimer` inside the onComplete of `removeFromMeetingRequestSent`

Reply:
Two thinks : 
First : I agree that there may be a racing condition here but i have an other proposition : doing 
removeFromMeetingRequestSent(senderUid) {removeFromMeetingRequestInbox(senderUid) .....}
Second : I think that the Sent onComplete function should just be all  the following lines, isn't it ?

Reply:
Yes that's exactly what I meant, it is perfect!